In [14]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.model_selection import StratifiedKFold, cross_val_score

In [15]:
X_train = pd.read_csv('../data/processed/X_train.csv')
y_train = pd.read_csv('../data/processed/y_train.csv')
X_test = pd.read_csv('../data/processed/X_test.csv')
y_test = pd.read_csv('../data/processed/y_test.csv')
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)


(76470, 12) (76470, 1)
(19118, 12) (19118, 1)


In [16]:
y_train.value_counts()

Label           
BENIGN              58943
DoS Hulk             6546
PortScan             5509
DDoS                 4625
DoS GoldenEye         372
DoS slowloris         209
DoS Slowhttptest      193
Bot                    71
Infiltration            2
Name: count, dtype: int64

In [17]:
print(X_train.dtypes)

Flow Duration             float64
Flow Bytes/s              float64
Flow Packets/s            float64
Total Fwd Packets         float64
Total Backward Packets    float64
Packet Length Mean        float64
Packet Length Std         float64
Average Packet Size       float64
Flow IAT Mean             float64
Flow IAT Std              float64
SYN Flag Count            float64
ACK Flag Count            float64
dtype: object


In [18]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [ ]:
lr_model = LogisticRegression(
    C = 1,
    solver = "lbfgs",
    max_iter = 1000,
)
lr_scores = cross_val_score(
    lr_model,
    X_train,
    y_train,
    cv=skf,
    scoring="accuracy"
)
print(f"Độ chính xác từng fold: {lr_scores}")
print(f"Độ chính xác trung bình: {lr_scores.mean()}")
print(f"Độ lệch chuẩn: {lr_scores.std()}")

d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\model_selection\_split.py:812: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.
  warnings.warn(
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\utils\validation.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\utils\validation.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\utils\validation.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to

Độ chính xác từng fold: [0.90708775 0.90519158 0.90323002 0.90231463 0.90355695]
Độ chính xác trung bình: 0.904276186739898
Độ lệch chuẩn: 0.0016853640506183964


In [20]:
svm_model = LinearSVC(
    C=1,
    random_state=42
)

svm_scores = cross_val_score(
    svm_model,
    X_train,
    y_train,
    cv=skf,
    scoring="accuracy"
)
print(f"Độ chính xác từng fold: {svm_scores}")
print(f"Độ chính xác trung bình: {svm_scores.mean()}")
print(f"Độ lệch chuẩn: {svm_scores.std()}")

d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\model_selection\_split.py:812: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.
  warnings.warn(
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\utils\validation.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\utils\validation.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\utils\validation.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to

Độ chính xác từng fold: [0.85334118 0.85811429 0.85634889 0.85641428 0.85889891]
Độ chính xác trung bình: 0.8566235124885576
Độ lệch chuẩn: 0.0019127354435441418


In [21]:
rf_model = RandomForestClassifier(
    random_state=42, 
    n_estimators=100, 
    max_depth=None, 
    oob_score=True, 
    n_jobs=-1
    )
rf_scores = cross_val_score(
    rf_model,
    X_train,
    y_train,
    cv=skf,
    scoring="accuracy"
)
print(f"Độ chính xác từng fold: {rf_scores}")
print(f"Độ chính xác trung bình: {rf_scores.mean()}")
print(f"Độ lệch chuẩn: {rf_scores.std()}")

d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\model_selection\_split.py:812: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.
  warnings.warn(
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\base.py:1403: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\base.py:1403: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)
d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\base.py:1403: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samp

Độ chính xác từng fold: [0.99130378 0.99365764 0.99241533 0.99182686 0.99248071]
Độ chính xác trung bình: 0.9923368641297241
Độ lệch chuẩn: 0.0007872324803605448


Dựa vào độ chính xác trung bình và độ lệch chuẩn của từng mô hình, ta thấy Random Forest là mô hình có độ chính xác cao nhất (99.23%) với độ lệch chuẩn là 0.00079. Dựa vào kết quả đó, nhóm chọn mô hình Random Forest để huấn luyện bộ dữ liệu

In [22]:
rf_model = RandomForestClassifier(
    random_state=42, 
    n_estimators=100, 
    max_depth=None, 
    oob_score=True, 
    n_jobs=-1)
rf_model.fit(X_train, y_train)

print(f"-> OOB Score (Ước lượng tập Train): {rf_model.oob_score_:.4f}")

y_test_pred = rf_model.predict(X_test)
test_acc = accuracy_score(y_test, y_test_pred)

print("\n--- ĐÁNH GIÁ TRÊN TẬP TEST ---")
print(classification_report(y_test, y_test_pred))
print(f"Accuracy trên tập Test: {test_acc:.4f}")

d:\bao_mat\BTL_Bao_mat_ung_dung_he_thong\venv\Lib\site-packages\sklearn\base.py:1403: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


-> OOB Score (Ước lượng tập Train): 0.9926

--- ĐÁNH GIÁ TRÊN TẬP TEST ---
                  precision    recall  f1-score   support

          BENIGN       1.00      1.00      1.00     14736
             Bot       0.86      0.33      0.48        18
            DDoS       1.00      0.99      0.99      1156
   DoS GoldenEye       0.98      0.97      0.97        93
        DoS Hulk       0.97      0.97      0.97      1637
DoS Slowhttptest       0.96      0.98      0.97        48
   DoS slowloris       1.00      0.92      0.96        52
        PortScan       0.99      1.00      1.00      1378

        accuracy                           0.99     19118
       macro avg       0.97      0.90      0.92     19118
    weighted avg       0.99      0.99      0.99     19118

Accuracy trên tập Test: 0.9932
